File: 1. Two Pointers/15. 3Sum/Solution.cpp
#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> threeSum(std::vector<int>& nums) {
        std::sort(nums.begin(), nums.end());
        std::vector<std::vector<int>> res {};

        for (int i = 0; i < nums.size(); ++i)
        {
            int a = nums[i];

            if (i > 0 && a == nums[i - 1])
            {
                continue;
            }

            if (a > 0)
            {
                break;
            }

            std::vector<std::vector<int>> pairs = twoSum(nums, i + 1, -a);

            for (auto& pair: pairs)
            {
                res.push_back({a, pair[0], pair[1]});
            }
        }

        return res;

    }

    std::vector<std::vector<int>> twoSum(std::vector<int>& nums, int left, int target)
    {
        int right = nums.size() - 1;
        std::vector<std::vector<int>> pairs {};

        while (left < right)
        {
            int sum = nums[left] + nums[right];

            if (sum == target)
            {
                pairs.push_back({nums[left], nums[right]});
                ++left;

                while (left < right && nums[left] == nums[left - 1])
                {
                    ++left;
                }
            }

            else if (sum < target)
            {
                ++left;
            }

            else
            {
                --right;
            }

        }

        return pairs;
    }
};

File: 1. Two Pointers/15. 3Sum/Solution.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        res = []

        for i in range(len(nums)):
            a = nums[i]

            if i > 0 and a == nums[i - 1]:
                continue
            
            if a > 0:
                break
            
            pairs = self.twoSum(nums, i + 1, -a)

            for pair in pairs:
                res.append([a] + pair)
        
        return res
    
    def twoSum(self, nums: list[int], left: int, target: int):
        right = len(nums) - 1
        pairs = []

        while left < right:
            sum = nums[left] + nums[right]

            if sum == target:
                pairs.append([nums[left], nums[right]])
                left += 1

                while left < right and nums[left] == nums[left - 1]:
                    left += 1


            elif sum < target:
                left += 1
            else:
                right -= 1
            
        return pairs
            